code/service/src/main/scala/com/namely/chiefofstate/AggregateRoot.scala

```diff
@@ -178,40 +178,55 @@ object AggregateRoot {
       protosValidator: ProtosValidator,
       data: Map[String, com.google.protobuf.any.Any]): ReplyEffect[EventWrapper, StateWrapper] = {
 
-    val handlerOutput: Try[WriteHandlerHelpers.WriteTransitions] = commandHandler
-      .handleCommand(command, priorState)
-      .map(_.event match {
-        case Some(newEvent) =>
-          protosValidator.requireValidEvent(newEvent)
-          WriteHandlerHelpers.NewEvent(newEvent)
-
-        case None =>
-          WriteHandlerHelpers.NoOp
-      })
-      .flatMap({
-        case WriteHandlerHelpers.NewEvent(newEvent) =>
+    // state will be updated with the resulting state of each HandleEventResponse
+    var state = priorState.getState
+
+    val handlerOutput: Try[WriteHandlerHelpers.WriteTransitions] =
+      commandHandler
+        .handleCommand(command, priorState)
+	// use events or fallback to event in case events is empty
+        .map(response =>
+          response.events.isEmpty match {
+            case true =>
+              response.event match {
+                case Some(newEvent) => Seq(newEvent)
+                case None           => Seq.empty[any.Any]
+              }
+            case _ => response.events
+          })
+        .map(_.filter(x => x != any.Any()) // ignore empty events and validate types
+          .map(newEvent => {
+            protosValidator.requireValidEvent(newEvent)
+            newEvent
+          }))
+	// call handleEvent for each event, while passing the updated resulting state from previous call
+        .map(_.map(newEvent => {
           val newEventMeta: MetaData = MetaData()
             .withRevisionNumber(priorState.getMeta.revisionNumber + 1)
             .withRevisionDate(Instant.now().toTimestamp)
             .withData(data)
             .withEntityId(priorState.getMeta.entityId)
             .withHeaders(command.persistedHeaders)
 
-          val priorStateAny: com.google.protobuf.any.Any = priorState.getState
+          val priorStateAny: com.google.protobuf.any.Any = state
 
           eventHandler
             .handleEvent(newEvent, priorStateAny, newEventMeta)
             .map(response => {
               require(response.resultingState.isDefined, "event handler replied with empty state")
               protosValidator.requireValidState(response.getResultingState)
-              WriteHandlerHelpers.NewState(newEvent, response.getResultingState, newEventMeta)
+              state = response.getResultingState
+              WriteHandlerHelpers.NewState(newEvent, state, newEventMeta)
             })
-
-        case x =>
-          Success(x)
-      })
-      .recoverWith(makeFailedStatusPf)
-
+        })
+	// extract WriteTransitions of each Try or fail, then get the last one (if any exists)
+	.map(_.get).lastOption)
```

Comment:
```suggestion
	.map(_.get).lastOption
      )

```

Reply:
wasn't that supposed to get auto formatted or something?

Reply:
I don't believe there's an auto format option for where parenthesis end. I just figure it would be much easier to understand this snippet since the `)` has nothing to do with that line.